futuredsp/src/firdes.rs

```diff
@@ -1,109 +1,106 @@
 //! Methods for designing FIR filters.
 
 extern crate alloc;
-use crate::windows::FilterWindow;
 use alloc::vec::Vec;
+use num_traits::FromPrimitive;
 
 /// Constructs a lowpass FIR filter with unit gain and cutoff frequency `cutoff` (in cycles/sample)
-/// using the specified window.
+/// using the specified window. The length of the filter equals the length of `window`.
+/// The filter taps are constructed internally as `f64` and then casted to the generic type `T`
+/// using [`num_traits::FromPrimitive::from_f64()`].
 ///
 /// Example usage:
 /// ```
 /// use futuredsp::{firdes, windows};
 ///
 /// let sampling_freq = 10_000;
 /// // 2000 Hz cutoff frequency, rectangular window
-/// let cutoff = 2_000.0 / sampling_freq as f32;
+/// let cutoff = 2_000.0 / sampling_freq as f64;
 /// let num_taps = 65;
-/// let rect_win = windows::RectWindow::new(num_taps);
-/// let taps = firdes::lowpass(num_taps, cutoff, rect_win);
+/// let rect_win = windows::rect(num_taps);
+/// let taps = firdes::lowpass::<f32>(cutoff, rect_win.as_slice());
 /// ```
-pub fn lowpass<T: FilterWindow<f32>>(num_taps: usize, cutoff: f32, window: T) -> Vec<f32> {
-    assert!(num_taps > 0, "num_taps must be greater than 0");
+pub fn lowpass<T: FromPrimitive>(cutoff: f64, window: &[f64]) -> Vec<T> {
     assert!(
         cutoff > 0.0 && cutoff < 1.0 / 2.0,
         "cutoff must be in (0, 1/2)"
     );
-    let mut taps: Vec<f32> = Vec::with_capacity(num_taps);
-    let omega_c = 2.0 * core::f32::consts::PI * cutoff;
-    let alpha = (num_taps - 1) as f32 / 2.0;
-    for n in 0..num_taps {
-        let x = n as f32 - alpha;
-        let tap = match x == 0.0 {
-            true => 1.0,
-            false => (omega_c * x).sin() / (core::f32::consts::PI * x),
-        };
-        taps.push(tap * window.get(n));
-    }
-
-    taps
+    let omega_c = 2.0 * core::f64::consts::PI * cutoff;
+    let alpha = (window.len() - 1) as f64 / 2.0;
+    window
+        .iter()
+        .enumerate()
+        .map(|(n, tap)| {
+            let x = n as f64 - alpha;
+            let filter_tap = match x == 0.0 {
+                true => 1.0,
+                false => (omega_c * x).sin() / (core::f64::consts::PI * x),
```

Comment:
Okay, so this is unrelated to this PR, but since we're here...

I made a simple flowgraph like this:
```rust
fn main() -> Result<()> {
    let mut fg = Flowgraph::new();

    let src = fg.add_block(futuresdr::blocks::Source::<Complex<f32>>::new(|| {
        let mut rng = rand::thread_rng();
        let x: f32 = rng.gen();
        let y: f32 = rng.gen();
        Complex {
            re: x * 2.0 - 1.0,
            im: y * 2.0 - 1.0,
        }
    }));

    let window = futuredsp::windows::BlackmanWindow::new(59);
    let taps = futuredsp::firdes::lowpass(59, 0.25, window);
    let filter = fg.add_block(FirBuilder::new::<Complex<f32>, f32, _>(taps));

    let sink = fg.add_block(FileSink::<Complex<f32>>::new("tmp.cf32"));

    fg.connect_stream(src, "out", filter, "in")?;
    fg.connect_stream(filter, "out", sink, "in")?;

    Runtime::new().run(fg)?;

    Ok(())
}
```
which, effectively, generates some complex white noise, runs it through a filter, and pipes that to a file which we can then analyze.

So, if I run this flowgraph as-is, I get a spectrum that looks something like:
![image](https://user-images.githubusercontent.com/879600/160287024-3fa9b7f9-8687-41af-bac1-85ca3691f67e.png)
with ~15dB of attenuation outside the passband.

If I change `omega_c` in `lowpass()` to `2.0 * cutoff` (removing the π factor) and change the "false" tap calculation to `(omega_c * x).sin() / (omega_c * x)` (from `(omega_c * x).sin() / (core::f32::consts::PI * x)`), so that the overall calculation looks like:
```rust
    let omega_c = 2.0 * cutoff;
    let alpha = (num_taps - 1) as f32 / 2.0;
    for n in 0..num_taps {
        let x = n as f32 - alpha;
        let tap = match x == 0.0 {
            true => 1.0,
            false => (omega_c * x).sin() / (omega_c * x),
        };
        taps.push(tap * window.get(n));
    }
```
I get:
![image](https://user-images.githubusercontent.com/879600/160287185-869b6074-d618-48ab-b196-300faac8b11f.png)
Obviously, the passband is a different width, but more notably the attenuation is higher outside the passband.

I know essentially nothing about signal processing (I stole my modifications from [here](https://tomroelandts.com/articles/how-to-create-a-simple-low-pass-filter)). I also haven't used the `firdes::kaiser` methods yet. But, is this known/expected behaviour?

Reply:
You are right that this is not how it is supposed to look.
I had a closer look, at it seems that the error is that the "true" tap should not be `1.0` but `omega_c / core::f32::consts::PI`. Changing that seems to solve the problem for me. I will provide a fix.

In fact, the implementation that we use is the same as the sinc filter in the link you provided:
```
h[n] = 2*fc*sinc(2*fc*n)
     = 2fc*sin(pi*2*fc*n) / (pi*2*fc*n)
     = sin(pi*2*fc*n) / (pi*n)
     = sin(omega_c*n) / (pi*n)
```
where `omega_c = pi*2*fc`. Since `sinc(0) = 1` we have to use `h[0] = 2*fc = omega_c / pi` (and obviously not `h[0] = 1.0`).

Reply:
Ah okay, that derivation makes sense to me. But yeah replacing the "true" tap worked on my end too, thanks!